### PR TITLE
BUG: Fix bug in pd.eval when UnaryOp in function

### LINKED ISSUE #46471: BUG: `.fillna(-1)` does not work in `pd.eval`
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
def test_unary(val):
    df = pd.DataFrame({'x': np.array([0.11, 0], dtype=np.float32)})
    res = df.eval(f'x=x.fillna({val})')
test_unary(1)  # works
test_unary(-1)  # fails
```

### Issue Description

`.eval` seems to be having some problems with negative values. A very similar issue has been reported, summarized and closed here: #16363 

However, it clearly still persists in other instances. 

### Expected Behavior

test_unary(-1)  # works

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit           : 66e3805b8cabe977f40c05259cc3fcf7ead5687d
python           : 3.9.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.13.0-35-generic
Version          : #40~20.04.1-Ubuntu SMP Mon Mar 7 09:18:32 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8
pandas           : 1.3.5
numpy            : 1.21.2
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 58.2.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.2
IPython          : 7.28.0
pandas_datareader: None
bs4              : None
bottleneck       : None
fsspec           : 2021.11.1
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 5.0.0
pyxlsb           : None
s3fs             : 2021.11.1
scipy            : 1.7.1
sqlalchemy       : 1.4.25
tables           : None
tabulate         : 0.8.9
xarray           : None
xlrd             : None
xlwt             : None
numba            : 0.55.0

</details>

### PR DESCRIPTION
**Pull request by yuanx749**
- [x] closes #46471
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

I ran into a similar issue, and added it in test. They both produce `AttributeError: 'UnaryOp' object has no attribute 'value'`.

### CODE REVIEW

File: pandas/tests/computation/test_eval.py [hunk 1]
```diff
@@ -600,6 +600,18 @@ def test_float_comparison_bin_op(self, dtype, expr):
         res = df.eval(expr)
         assert res.values == np.array([False])
 
+    def test_unary_in_function(self):
+        # GH 46471
+        df = DataFrame({"x": [0, 1, np.nan]})
+
+        result = df.eval("x.fillna(-1)")
+        expected = df.x.fillna(-1)
+        tm.assert_series_equal(result, expected, check_names=False)
```
Comment (mroeschke) [inline pandas/tests/computation/test_eval.py:609]: Why `check_names=False`?

**Review by mroeschke — COMMENTED** [review]
Could you add a whatsnew not in `1.6.0.rst` (in `Conversion` section)?

File: pandas/tests/computation/test_eval.py (hunk 1 above)
Comment (yuanx749) [reply pandas/tests/computation/test_eval.py:609]: Weird, the name check passed locally, but failed on not single cpu and on CircleCI.

See the results of my previous commit [here](https://github.com/pandas-dev/pandas/actions/runs/3034645554/jobs/4883948812).

**Comment by yuanx749** [comment]
> Could you add a whatsnew not in `1.6.0.rst` (in `Conversion` section)?

Added.

File: pandas/tests/computation/test_eval.py (hunk 1 above)
Comment (yuanx749) [reply pandas/tests/computation/test_eval.py:609]: > Weird, the name check passed locally, but failed on not single cpu and on CircleCI.
> 
> See the results of my previous commit [here](https://github.com/pandas-dev/pandas/actions/runs/3034645554/jobs/4883948812).

Seems it's caused by numexpr engine. After installed locally, the column name becomes `None` using `df.eval`.
I noticed `check_names=False` appears several times in `test_eval.py`.

Comment (mroeschke) [reply pandas/tests/computation/test_eval.py:609]: Could we make this check more specific then to only check names when the engine is not numexpr?

Comment (yuanx749) [reply pandas/tests/computation/test_eval.py]: Changed to `check_names=not USE_NUMEXPR`. Do you think it is OK?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @yuanx749

**Pull request merged**